core/src/main/java/io/questdb/griffin/engine/functions/str/SplitPartFunctionFactory.java

```diff
@@ -0,0 +1,210 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.str;
+
+import io.questdb.cairo.CairoConfiguration;
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.functions.BinaryFunction;
+import io.questdb.griffin.engine.functions.StrFunction;
+import io.questdb.griffin.engine.functions.TernaryFunction;
+import io.questdb.griffin.engine.functions.constants.StrConstant;
+import io.questdb.std.*;
+import io.questdb.std.str.CharSink;
+import io.questdb.std.str.StringSink;
+import org.jetbrains.annotations.Nullable;
+
+public class SplitPartFunctionFactory implements FunctionFactory {
+    @Override
+    public String getSignature() {
+        return "split_part(SI)";
+    }
+
+    @Override
+    public Function newInstance(int position, ObjList<Function> args, IntList argPositions, CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) throws SqlException {
+        final Function strFunc = args.getQuick(0);
+        final Function delimiterFunc = args.getQuick(1);
```

Comment:
nit: would be great to have separate implementations for the single char delimiter case since it's very likely to be a common scenario among our users. Having a `char` instead of a `j.l.CharSequence` as the delimiter, would simplify the lookup loop and reduce the number of branches yielding better performance.

Reply:
How can I test the char implementation with `assertQuery()`? I tried calling it with:
```java
assertQuery(
                "split_part\n" + "def\n",
                "select split_part('abc,def,ghi,jkl', ',', 2)",
                null,
                true,
                true);
```
but it calls the `CharSequence` implementation instead

Reply:
> I just realized that the single character String will resolve to a `char` while a multi-character String will resolve to a `CharSequence`.

Yes, that's correct.

Reply:
Because the char delimiter arg function can be a constant or dynamic function, and the index arg function can be a constant or runtime constant function, the number of function implementations greatly increases, especially if we want to also optimize for the char delimiter arg function being a runtime constant function.

To reduce the amount of code duplication, I've abstracted out the arg function handling logic to a new utility class called `ArgFunctions`.